Draft release notes for 1.64 (#1027)
rust-lang/blog.rust-lang.org: posts/2022-09-22-Rust-1.64.0.md

considered a stable interface. Nonetheless, because multiple crates relied on
this, we first worked with the authors of all of the still-maintained crates
doing so to arrange for the release of new versions, and to yank the old ones,
before making this change.

See <https://github.com/rust-lang/rust/pull/78802> for more details.

### Warning about unused tuple-struct fields

If you define a struct containing fields that are never used, rustc will warn
about the unused fields. Now, in Rust 1.64, you can enable the
`unused_tuple_struct_fields` lint to get the same warnings about unused fields
in a tuple struct. In future versions, we plan to make this lint warn by
default.

Fields of type unit (`()`) do not produce a warning, to make it easier to
migrate existing code without having to change tuple indices.

### Enhancing `.await` with `IntoFuture`

Rust 1.64 stabilizes the
[`IntoFuture`](https://doc.rust-lang.org/std/future/trait.IntoFuture.html)
trait. `IntoFuture` is a trait similar to
[`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html),
but rather than supporting `for ... in ...` loops, `IntoFuture` changes how

> `.await` works. With `IntoFuture`, the `.await` keyword can await mor ethan

Reviewer: ```suggestion
`.await` works. With `IntoFuture`, the `.await` keyword can await more than
```

    pub fn new() -> Self { ... }
    /// Decide whether debug mode should be enabled.
    pub fn set_debug(self, b: bool) -> Self { ... }
    /// Send the request and receive a response.
    pub async fn send(self) -> Result<StorageResponse, Error> { ... }
}

// The new implementations:
// 1. create a new named future type
// 2. implement `IntoFuture` for `StorageRequest`
pub type StorageRequestFuture = Pin<Box<dyn Future<Output = Result<StorageResponse, Error> + Send + 'static>>
impl IntoFuture for StorageRequest {
    type IntoFuture = StorageRequestFuture;
    type Output = <StorageRequestFuture as Future>::Output;
    fn into_future(self) -> Self::IntoFuture {
        Box::pin(self.send())
    }
}
```
This takes a bit more code to implement, but provides a simpler API for users.

In the future, the Rust Async WG hopes to simplify the creating new named
futures by supporting [`impl Trait` in `type` aliases (Type Alias Impl Trait or
TAIT)](https://rust-lang.github.io/impl-trait-initiative/explainer/tait.html).
This should make implementing `IntoFuture` easier by simplifying the type
alias' signature, and make it more performant by removing the `Box` from the
type alias.

> ### Stabilizing `ready!` and `poll_fn`

Reviewer: Something about formatting went wrong -- not quite sure where -- but probably worth checking the generated content even if github's editor highlighting isn't fixed.

Author: The generated content looks correct, and vim's highlighting seems fine with it too. This seems like a bug in github's highlighting.

---
layout: post
title: "Announcing Rust 1.64.0"
author: Josh Triplett

> team: The Release Team <https://www.rust-lang.org/governance/teams/release>

Reviewer 2: Nit: we usually just sign these posts as the release team regardless of who writes them.

If you'd like to help us out by testing future releases, you might consider
updating locally to use the beta channel (`rustup default beta`) or the nightly
channel (`rustup default nightly`). Please
[report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you
might come across!

## What's in 1.64.0 stable

### C-compatible FFI types in core and alloc

When calling or being called by C ABIs, Rust code can use type aliases like
`c_uint` or `c_ulong` to match the corresponding types from C on any target,
without requiring target-specific code or conditionals.

Previously, these type aliases were only available in `std`, so code written
for embedded targets and other scenarios that could only use `core` or `alloc`
could not use these types.

Rust 1.64 now provides all of the `c_*` type aliases in
[`core::ffi`](https://doc.rust-lang.org/core/ffi/index.html), as well as
[`core::ffi::CStr`](https://doc.rust-lang.org/core/ffi/struct.CStr.html) for
working with C strings. Rust 1.64 also provides
[`alloc::ffi::CString`](https://doc.rust-lang.org/alloc/ffi/struct.CString.html)
for working with owned C strings using only the `alloc` crate, rather than the
full `std` library.

> ### New internal layout for network structures

Reviewer 2: I would move this down the post: it's in the blog mostly for compatibility, not because it has a high user-facing impact.

  returns the contained value in a `Poll::Ready`. This can be useful when
  implementing futures manually using `Poll`.
- [`poll_fn`](https://doc.rust-lang.org/std/future/fn.poll_fn.html) turns a
  function returning `Poll` into a future.

In the future, we hope to provide simpler APIs that require less use of
low-level details like `Poll` and `Pin`, but in the meantime, these helpers
make it easier to write such code.

### Minimum required kernel and glibc versions increased

As [previously
announced](https://blog.rust-lang.org/2022/08/01/Increasing-glibc-kernel-requirements.html),
`linux` targets now require at least Linux kernel 3.2 (except for targets which
already required a newer kernel), and `linux-gnu` targets now require glibc
2.17 (except for targets which already required a newer glibc).

### Profile-guided optimization (PGO) for Windows targets

Windows builds of the Rust compiler now use profile-guided optimization,
providing performance improvements of 10-20% for compiling Rust code on
Windows.

### `rust-analyzer` now available via `rustup`, no longer in preview

Users of `rust-analyzer` can now obtain it via `rustup` on the stable and beta

> channels, not just on nightly.

Reviewer 2: This seems like a big news to me, and we should make it more prominent and write a tiny more, possibly including the command to add rust-analyzer with rustup.

Reviewer 3: A caveat is that they _don't_ want VS Code users to go that way, as the extension grabs the weekly rust-analyzer builds itself.

Reviewer 4: I was hoping to have a new rustup release ready before 1.64. As it stands now, it is a bit awkward to use.  The user needs to run `rustup which rust-analyzer` to get the path to the executable to run.  (With a new release, it would just be `rust-analyzer`).

I would encourage adding some text here to clarify what this means or what the consequences are. Maybe something along the lines of:

> [rust-analyzer](https://rust-analyzer.github.io/) is now included as part of the collection of tools included with Rust. This can make it easier to download and access rust-analyzer, and makes it available on more platforms. It is available as a [rustup component](https://rust-lang.github.io/rustup/concepts/components.html) which can be installed with `rustup component add rust-analyzer`. 
> 
> At this time, to run the rustup-installed version, use the command `rustup which rust-analyzer` to obtain the path to the executable. The next release of rustup will have a built-in proxy so that running the executable `rust-analyzer` will launch the appropriate version.
> 
> Most users should continue to use the releases provided by the rust-analyzer team (available on the [rust-analyzer releases page](https://github.com/rust-lang/rust-analyzer/releases)), which are published more frequently. Users of the [official VSCode extension](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer) are not affected since it automatically downloads new releases.

However, after writing that, it seems like way too many words for something that I think is relatively minor.

Reviewer 2: Note that there is also `rustup run rust-analyzer` which works, even though it's not a singl ebinary.

Cargo workspace, you can now avoid duplication of common field values between
crates, such as common version numbers, repository URLs, or `rust-version`.
This also helps keep these values in sync between crates when updating them.
For more details, see
[`workspace.package`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspacepackage-table),
[`workspace.dependencies`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspacedependencies-table),
and ["inheriting a dependency from a
workspace"](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inheriting-a-dependency-from-a-workspace).

### Cargo multi-target builds

When building for multiple targets, you can now pass multiple `--target`
options to `cargo build`, to build all of those targets at once. You can also
set [`build.target`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget) to an array of multiple targets in `.cargo/config.toml` to
build for multiple targets by default.

### Stabilized APIs

The following methods and trait implementations are now stabilized:

- ...

These APIs are now usable in const contexts:

> - ...

Reviewer 2: This is missing.

- [`ffi::c_double`](https://doc.rust-lang.org/stable/std/ffi/type.c_double.html)
- [`ffi::c_float`](https://doc.rust-lang.org/stable/std/ffi/type.c_float.html)
- [`ffi::c_int`](https://doc.rust-lang.org/stable/std/ffi/type.c_int.html)
- [`ffi::c_long`](https://doc.rust-lang.org/stable/std/ffi/type.c_long.html)
- [`ffi::c_longlong`](https://doc.rust-lang.org/stable/std/ffi/type.c_longlong.html)
- [`ffi::c_schar`](https://doc.rust-lang.org/stable/std/ffi/type.c_schar.html)
- [`ffi::c_short`](https://doc.rust-lang.org/stable/std/ffi/type.c_short.html)
- [`ffi::c_uchar`](https://doc.rust-lang.org/stable/std/ffi/type.c_uchar.html)
- [`ffi::c_uint`](https://doc.rust-lang.org/stable/std/ffi/type.c_uint.html)
- [`ffi::c_ulong`](https://doc.rust-lang.org/stable/std/ffi/type.c_ulong.html)
- [`ffi::c_ulonglong`](https://doc.rust-lang.org/stable/std/ffi/type.c_ulonglong.html)
- [`ffi::c_ushort`](https://doc.rust-lang.org/stable/std/ffi/type.c_ushort.html)

These APIs are now usable in const contexts:

- [`slice::from_raw_parts`](https://doc.rust-lang.org/stable/core/slice/fn.from_raw_parts.html)

### Compatibility notes

> Rust 1.64 includes some changes that might prevent you from upgrading:
> 

Author: ```suggestion
```

is also the last release containing a copy of RLS. Starting from Rust 1.65.0,
RLS will be replaced by a small LSP server showing the deprecation warning.

### Other changes

There are other changes in the Rust 1.64 release, including:

* Windows builds of the Rust compiler now use profile-guided optimization,
  providing performance improvements of 10-20% for compiling Rust code on
  Windows.

* If you define a struct containing fields that are never used, rustc will warn
  about the unused fields. Now, in Rust 1.64, you can enable the
  `unused_tuple_struct_fields` lint to get the same warnings about unused
  fields in a tuple struct. In future versions, we plan to make this lint
  warn by default.  Fields of type unit (`()`) do not produce this warning,
  to make it easier to migrate existing code without having to change tuple
  indices.

* When working with collections of related libraries or binary crates in one
  Cargo workspace, you can now avoid duplication of common field values between
  crates, such as common version numbers, repository URLs, or `rust-version`.
  This also helps keep these values in sync between crates when updating them.
  For more details, see

>   [`workspace.package`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspacepackage-table),
>   [`workspace.dependencies`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspacedependencies-table),
>   and ["inheriting a dependency from a
>   workspace"](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inheriting-a-dependency-from-a-workspace).

Reviewer 5: I believe [this beta backport](https://github.com/rust-lang/cargo/pull/11097) changed the links around. The change isn't showing in the [beta docs](https://doc.rust-lang.org/beta/cargo/reference/workspaces.html) but they are showing in the [nightly docs](https://doc.rust-lang.org/nightly/cargo/reference/workspaces.html)
```suggestion
  [`workspace.package`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-package-table),
  [`workspace.dependencies`](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table),
  and ["inheriting a dependency from a
  workspace"](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inheriting-a-dependency-from-a-workspace).
```

- ...

### Compatibility notes

> Rust 1.64 includes some changes that might prevent you from upgrading:

Author: Let's not use the phrasing "prevent you from upgrading". I think these two changes can just stand on their own without a lead-in.

Reviewer 2: Good call.
